src/Microsoft.IdentityModel.Tokens/ECDsaAdapter.cs

```diff
@@ -0,0 +1,401 @@
+        private ECDsa CreateECDsaUsingECParams(JsonWebKey jsonWebKey, bool usePrivateKey)
+        {
+            if (jsonWebKey == null)
+                throw LogHelper.LogArgumentNullException(nameof(jsonWebKey));
+
+            if (jsonWebKey.Crv == null)
+                throw LogHelper.LogArgumentNullException(nameof(jsonWebKey.Crv));
+
+            if (jsonWebKey.X == null)
+                throw LogHelper.LogArgumentNullException(nameof(jsonWebKey.X));
+
+            if (jsonWebKey.Y == null)
+                throw LogHelper.LogArgumentNullException(nameof(jsonWebKey.Y));
+
+            try
+            {
+                var ecParams = new ECParameters
+                {
+                    Curve = GetNamedECCurve(jsonWebKey.Crv),
+                    Q = {X = Base64UrlEncoder.DecodeBytes(jsonWebKey.X), Y = Base64UrlEncoder.DecodeBytes(jsonWebKey.Y)}
+                };
+
+                if (usePrivateKey)
+                {
+                    if (jsonWebKey.D == null)
+                        throw LogHelper.LogArgumentNullException(nameof(jsonWebKey.D));
+
+                    ecParams.D = Base64UrlEncoder.DecodeBytes(jsonWebKey.D);
+                }
+
+                return ECDsa.Create(ecParams);
+            }
+            catch (Exception ex)
+            {
+                throw LogHelper.LogExceptionMessage(new CryptographicException(LogMessages.IDX10689, ex));
+            }
+        }
+
+        /// <summary>
+        /// Returns the elliptic curve corresponding to the curve id.
+        /// </summary>
+        /// <param name="curveId">Represents ecdsa curve -P256, P384, P512</param>
+        private ECCurve GetNamedECCurve(string curveId)
+        {
+            if (string.IsNullOrEmpty(curveId))
+                throw LogHelper.LogArgumentNullException(nameof(curveId));
+
+            switch (curveId)
+            {
+                case JsonWebKeyECTypes.P256:
+                    return ECCurve.NamedCurves.nistP256;
+                case JsonWebKeyECTypes.P384:
+                    return ECCurve.NamedCurves.nistP384;
+                case JsonWebKeyECTypes.P512: // treat 512 as 521. 512 doesn't exist, but we released with "512" instead of "521", so don't break now.
+                case JsonWebKeyECTypes.P521:
+                    return ECCurve.NamedCurves.nistP521;
+                default:
+                    throw LogHelper.LogExceptionMessage(new ArgumentException(LogHelper.FormatInvariant(LogMessages.IDX10645, curveId)));
+            }
+        }
+
+        /// <summary>
+        /// Tests if user application's runtime supports <see cref="ECParameters"/> structure.
+        /// </summary>
+        /// <returns>True if <see cref="ECParameters"/> structure is supported, false otherwise.</returns>
+        private bool SupportsECParameters()
+        {
+            try
+            {
+                LoadECParametersType();
+                return true;
+            }
+            catch (TypeLoadException)
```

Comment:
Would there be any harm in catching "Exception" ?

Reply:
I guess not. Catching any exception will make the code more resilient, as discussed. Addressed with https://github.com/AzureAD/azure-activedirectory-identitymodel-extensions-for-dotnet/pull/1058/commits/2ad287831df091cb053b771d931553565efec723